kubernetes/kubernetes_service.go

```diff
@@ -135,94 +140,87 @@ func (in *IstioClient) GetDeploymentConfigs(namespace string) ([]osappsv1.Deploy
-	sf, err := in.k8s.AppsV1beta2().StatefulSets(namespace).List(emptyListOptions)
-	if err != nil {
-		return nil, err
+	sf, found := in.k8sCache.GetStatefulSets(namespace)
+	if !found {
+		sf = []v1beta2.StatefulSet{}
 	}
-	return sf.Items, nil
-}
-
-func (in *IstioClient) GetReplicationController(namespace string, replicasetName string) (*v1.ReplicationController, error) {
-	return in.k8s.CoreV1().ReplicationControllers(namespace).Get(replicasetName, emptyGetOptions)
+	return sf, nil
 }
 
 func (in *IstioClient) GetReplicationControllers(namespace string) ([]v1.ReplicationController, error) {
-	rc, err := in.k8s.CoreV1().ReplicationControllers(namespace).List(emptyListOptions)
-	if err != nil {
-		return nil, err
+	rc, found := in.k8sCache.GetReplicationControllers(namespace)
+	if !found {
+		rc = []v1.ReplicationController{}
 	}
-	return rc.Items, nil
+	return rc, nil
 }
 
 // GetService returns the definition of a specific service.
 // It returns an error on any problem.
 func (in *IstioClient) GetService(namespace, serviceName string) (*v1.Service, error) {
-	return in.k8s.CoreV1().Services(namespace).Get(serviceName, emptyGetOptions)
+	if svc, exist := in.k8sCache.GetService(namespace, serviceName); exist {
+		return svc, nil
+	}
+	return nil, NewNotFound(serviceName, "core/v1", "Service")
 }
 
 // GetEndpoints return the list of endpoint of a specific service.
 // It returns an error on any problem.
 func (in *IstioClient) GetEndpoints(namespace, serviceName string) (*v1.Endpoints, error) {
-	return in.k8s.CoreV1().Endpoints(namespace).Get(serviceName, emptyGetOptions)
+	if endpoints, exist := in.k8sCache.GetEndpoints(namespace, serviceName); exist {
+		return endpoints, nil
+	}
+	return nil, NewNotFound(serviceName, "core/v1", "Endpoints")
 }
 
 // GetPods returns the pods definitions for a given set of labels.
 // An empty labelSelector will fetch all pods found per a namespace.
 // It returns an error on any problem.
 func (in *IstioClient) GetPods(namespace, labelSelector string) ([]v1.Pod, error) {
-	// An empty selector is ambiguous in the go client, could mean either "select all" or "select none"
-	// Here we assume empty == select all
-	// (see also https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors)
-	pods, err := in.k8s.CoreV1().Pods(namespace).List(meta_v1.ListOptions{LabelSelector: labelSelector})
-	if err != nil {
-		return nil, err
+	pods, found := in.k8sCache.GetPods(namespace)
+	if !found {
+		pods = []v1.Pod{}
 	}
-	return pods.Items, nil
-}
-
-// GetCronJob returns the definition of a specific CronJob.
-// It returns an error on any problem.
-func (in *IstioClient) GetCronJob(namespace, cronjobName string) (*batch_v1beta1.CronJob, error) {
-	return in.k8s.BatchV1beta1().CronJobs(namespace).Get(cronjobName, emptyGetOptions)
+	if labelSelector != "" {
+		selector, err := labels.Parse(labelSelector)
+		if err != nil {
+			return pods, err
```

Comment:
I think it would be safer to return an empty array here, to avoid the risk that the caller takes the list of pods ignoring the error.

Reply:
+1

Reply:
don't forget this one :)

Reply:
Oh, true thanks ! I'm on my way, having issues with the local test env this morning, but I'll update it